# Employee Performance & Retention Analysis

In [2]:
#🔹 STEP 0: Load the Dataset

In [7]:
import pandas as pd
df = pd.read_csv(r'D:\Projects\employees_dataset.csv')

In [8]:
df.head()

employee_id         department     region  education gender  \
0         8724         Technology  region_26  Bachelors      m   
1        74430                 HR   region_4  Bachelors      f   
2        72255  Sales & Marketing  region_13  Bachelors      m   
3        38562        Procurement   region_2  Bachelors      f   
4        64486            Finance  region_29  Bachelors      m   

  recruitment_channel  no_of_trainings  age  previous_year_rating  \
0            sourcing                1   24                   NaN   
1               other                1   31                   3.0   
2               other                1   31                   1.0   
3               other                3   31                   2.0   
4            sourcing                1   30                   4.0   

   length_of_service  KPIs_met_more_than_80  awards_won  avg_training_score  
0                  1                      1           0                  77  
1                  5                      0           0                  51  
2                  4                      0           0                  47  
3                  9                      0           0                  65  
4                  7                      0           0                  61

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17417 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17417 non-null  int64  
 1   department             17417 non-null  object 
 2   region                 17417 non-null  object 
 3   education              16646 non-null  object 
 4   gender                 17417 non-null  object 
 5   recruitment_channel    17417 non-null  object 
 6   no_of_trainings        17417 non-null  int64  
 7   age                    17417 non-null  int64  
 8   previous_year_rating   16054 non-null  float64
 9   length_of_service      17417 non-null  int64  
 10  KPIs_met_more_than_80  17417 non-null  int64  
 11  awards_won             17417 non-null  int64  
 12  avg_training_score     17417 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.7+ MB


In [10]:
df.shape

(17417, 13)

In [11]:
# Missing values exist only in background attributes, not core performance metrics.

In [13]:
# Percentage of missing data
df.isna().sum()

employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

In [14]:
(df.isna().sum() / len(df)) * 100

employee_id              0.000000
department               0.000000
region                   0.000000
education                4.426710
gender                   0.000000
recruitment_channel      0.000000
no_of_trainings          0.000000
age                      0.000000
previous_year_rating     7.825688
length_of_service        0.000000
KPIs_met_more_than_80    0.000000
awards_won               0.000000
avg_training_score       0.000000
dtype: float64

In [15]:
df.describe()

employee_id  no_of_trainings           age  previous_year_rating  \
count  17417.000000     17417.000000  17417.000000          16054.000000   
mean   39083.491129         1.250732     34.807774              3.345459   
std    22707.024087         0.595692      7.694046              1.265386   
min        3.000000         1.000000     20.000000              1.000000   
25%    19281.000000         1.000000     29.000000              3.000000   
50%    39122.000000         1.000000     33.000000              3.000000   
75%    58838.000000         1.000000     39.000000              4.000000   
max    78295.000000         9.000000     60.000000              5.000000   

       length_of_service  KPIs_met_more_than_80    awards_won  \
count       17417.000000           17417.000000  17417.000000   
mean            5.801860               0.358845      0.023368   
std             4.175533               0.479675      0.151074   
min             1.000000               0.000000      0.000000   
25%             3.000000               0.000000      0.000000   
50%             5.000000               0.000000      0.000000   
75%             7.000000               1.000000      0.000000   
max            34.000000               1.000000      1.000000   

       avg_training_score  
count        17417.000000  
mean            63.176322  
std             13.418179  
min             39.000000  
25%             51.000000  
50%             60.000000  
75%             75.000000  
max             99.000000

In [16]:
# Check categorical distributions
df.describe(include='object')

department    region  education gender recruitment_channel
count               17417     17417      16646  17417               17417
unique                  9        34          3      2                   3
top     Sales & Marketing  region_2  Bachelors      m               other
freq                 5458      3918      11519  12314                9751

## 1. Data Understanding and Initial Exploration

### The dataset contains 17,417 employee records with 13 features, including demographic, performance, and training-related variables. 
### Most columns are complete, with missing values observed only in education and previous_year_rating. 
### Data types are appropriate, and the dataset is ready for preprocessing and exploratory analysis.

## Exploratory Data Analysis (EDA) – Core Analysis

In [19]:
df.duplicated().sum()
# Duplicate employees = wrong retention & productivity insights

np.int64(2)

In [20]:
# The dataset contained 2 duplicate records, which were identified during initial data exploration.

In [21]:
df = df.drop_duplicates()

In [22]:
df

employee_id         department     region  education gender  \
0             8724         Technology  region_26  Bachelors      m   
1            74430                 HR   region_4  Bachelors      f   
2            72255  Sales & Marketing  region_13  Bachelors      m   
3            38562        Procurement   region_2  Bachelors      f   
4            64486            Finance  region_29  Bachelors      m   
...            ...                ...        ...        ...    ...   
17410        57239  Sales & Marketing  region_31  Bachelors      m   
17411        73858  Sales & Marketing  region_25  Bachelors      m   
17412        64573         Technology   region_7  Bachelors      f   
17413        49584                 HR   region_7  Bachelors      m   
17416        64573                 HR   region_7  Bachelors      f   

      recruitment_channel  no_of_trainings  age  previous_year_rating  \
0                sourcing                1   24                   NaN   
1                   other                1   31                   3.0   
2                   other                1   31                   1.0   
3                   other                3   31                   2.0   
4                sourcing                1   30                   4.0   
...                   ...              ...  ...                   ...   
17410               other                1   42                   3.0   
17411            sourcing                2   28                   2.0   
17412            referred                2   30                   5.0   
17413               other                1   33                   1.0   
17416            referred                1   35                   5.0   

       length_of_service  KPIs_met_more_than_80  awards_won  \
0                      1                      1           0   
1                      5                      0           0   
2                      4                      0           0   
3                      9                      0           0   
4                      7                      0           0   
...                  ...                    ...         ...   
17410                 10                      0           0   
17411                  5                      0           0   
17412                  6                      1           0   
17413                  9                      0           0   
17416                  4                      1           0   

       avg_training_score  
0                      77  
1                      51  
2                      47  
3                      65  
4                      61  
...                   ...  
17410                  53  
17411                  45  
17412                  81  
17413                  51  
17416                  81  

[17415 rows x 13 columns]

In [23]:
df.duplicated().sum() # 0 duplicates 

np.int64(0)

## 2. Data Cleaning and Preprocessing

### Handling Missing Values

In [25]:
df.isna().sum()

employee_id                 0
department                  0
region                      0
education                 771
gender                      0
recruitment_channel         0
no_of_trainings             0
age                         0
previous_year_rating     1363
length_of_service           0
KPIs_met_more_than_80       0
awards_won                  0
avg_training_score          0
dtype: int64

In [26]:
df['education'].mode() # to fill the missing eductaion with mode-here its bachelors

0    Bachelors
Name: education, dtype: object

In [27]:
df['education'] = df['education'].fillna(df['education'].mode()[0])

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17415 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17415 non-null  int64  
 1   department             17415 non-null  object 
 2   region                 17415 non-null  object 
 3   education              17415 non-null  object 
 4   gender                 17415 non-null  object 
 5   recruitment_channel    17415 non-null  object 
 6   no_of_trainings        17415 non-null  int64  
 7   age                    17415 non-null  int64  
 8   previous_year_rating   16052 non-null  float64
 9   length_of_service      17415 non-null  int64  
 10  KPIs_met_more_than_80  17415 non-null  int64  
 11  awards_won             17415 non-null  int64  
 12  avg_training_score     17415 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.9+ MB


In [30]:
# handle previous year rating
df['previous_year_rating'] = df['previous_year_rating'].fillna(
    df['previous_year_rating'].median()
)

In [31]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 17415 entries, 0 to 17416
Data columns (total 13 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   employee_id            17415 non-null  int64  
 1   department             17415 non-null  object 
 2   region                 17415 non-null  object 
 3   education              17415 non-null  object 
 4   gender                 17415 non-null  object 
 5   recruitment_channel    17415 non-null  object 
 6   no_of_trainings        17415 non-null  int64  
 7   age                    17415 non-null  int64  
 8   previous_year_rating   17415 non-null  float64
 9   length_of_service      17415 non-null  int64  
 10  KPIs_met_more_than_80  17415 non-null  int64  
 11  awards_won             17415 non-null  int64  
 12  avg_training_score     17415 non-null  int64  
dtypes: float64(1), int64(7), object(5)
memory usage: 1.9+ MB


In [32]:
df.isna().sum()

employee_id              0
department               0
region                   0
education                0
gender                   0
recruitment_channel      0
no_of_trainings          0
age                      0
previous_year_rating     0
length_of_service        0
KPIs_met_more_than_80    0
awards_won               0
avg_training_score       0
dtype: int64

### Missing values in the education and previous_year_rating columns were handled appropriately. The education column was imputed using the mode, while previous_year_rating was imputed using the median to maintain robustness against outliers. After imputation, the dataset contained no missing values and was ready for exploratory analysis.

# Step 3: Exploratory Data Analysis (EDA)

### EDA – Section 3.1 - Performance Patterns Analysis

In [34]:
# Objective - Identify performance patterns based on age, education, department, and training

In [36]:
# 3.1.1 Overall performance snapshot
df[['KPIs_met_more_than_80',
    'previous_year_rating',
    'avg_training_score',
    'awards_won']].describe()

KPIs_met_more_than_80  previous_year_rating  avg_training_score  \
count           17415.000000          17415.000000        17415.000000   
mean                0.358886              3.318691           63.177720   
std                 0.479687              1.218216           13.418314   
min                 0.000000              1.000000           39.000000   
25%                 0.000000              3.000000           51.000000   
50%                 0.000000              3.000000           60.000000   
75%                 1.000000              4.000000           75.000000   
max                 1.000000              5.000000           99.000000   

         awards_won  
count  17415.000000  
mean       0.023371  
std        0.151082  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        0.000000  
max        1.000000

In [37]:
overall_snapshot = {
    "Average Training Score": df['avg_training_score'].mean(),
    "Average Previous Year Rating": df['previous_year_rating'].mean(),
    "Average Number of Trainings": df['no_of_trainings'].mean(),
    "KPI Achievement Rate (%)": df['KPIs_met_more_than_80'].mean() * 100,
    "Award Winning Rate (%)": df['awards_won'].mean() * 100
}

overall_snapshot


{'Average Training Score': np.float64(63.17772035601493),
 'Average Previous Year Rating': np.float64(3.318690783807063),
 'Average Number of Trainings': np.float64(1.2507608383577375),
 'KPI Achievement Rate (%)': np.float64(35.888601780074644),
 'Award Winning Rate (%)': np.float64(2.3370657479184613)}

In [38]:
snapshot_df = pd.DataFrame(
    overall_snapshot.items(),
    columns=['Metric', 'Value']
)
snapshot_df

Metric      Value
0        Average Training Score  63.177720
1  Average Previous Year Rating   3.318691
2   Average Number of Trainings   1.250761
3      KPI Achievement Rate (%)  35.888602
4        Award Winning Rate (%)   2.337066

In [39]:
df.groupby('education')[['KPIs_met_more_than_80',
                          'avg_training_score',
                          'previous_year_rating']].mean().reset_index()


education  KPIs_met_more_than_80  avg_training_score  \
0        Bachelors               0.352702           62.737467   
1  Below Secondary               0.465035           65.902098   
2  Masters & above               0.368312           64.134270   

   previous_year_rating  
0              3.297933  
1              3.367133  
2              3.368519

In [40]:
# Does higher education correlate with performance?
# Training score differences

In [41]:
# 3.1.4 Performance by Age

In [44]:
# create age bands
df['age_group'] = pd.cut(
    df['age'],
    bins=[20, 30, 40, 50, 60],
    labels=['20–30', '31–40', '41–50', '51–60']
)

In [45]:
df.groupby('age_group')[['KPIs_met_more_than_80',
                          'avg_training_score',
                          'previous_year_rating']].mean().reset_index()

C:\Users\jaffe\AppData\Local\Temp\ipykernel_9268\791485626.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('age_group')[['KPIs_met_more_than_80',


age_group  KPIs_met_more_than_80  avg_training_score  previous_year_rating
0     20–30               0.364601           63.787330              3.245910
1     31–40               0.368357           63.155175              3.349629
2     41–50               0.331447           61.892078              3.357316
3     51–60               0.317204           63.254839              3.398925

In [46]:
df.groupby(
    'age_group',
    observed=False
)[['KPIs_met_more_than_80', 'avg_training_score']].mean()


KPIs_met_more_than_80  avg_training_score
age_group                                           
20–30                   0.364601           63.787330
31–40                   0.368357           63.155175
41–50                   0.331447           61.892078
51–60                   0.317204           63.254839

In [47]:
# 3.1.5 Performance vs Training Count

In [48]:
df.groupby('no_of_trainings')[['KPIs_met_more_than_80',
                               'avg_training_score']].mean().reset_index()


no_of_trainings  KPIs_met_more_than_80  avg_training_score
0                1               0.367736           62.789530
1                2               0.333333           65.026256
2                3               0.306122           64.220779
3                4               0.204380           62.678832
4                5               0.307692           66.576923
5                6               0.000000           62.250000
6                7               0.166667           63.666667
7                8               0.500000           71.000000
8                9               0.500000           70.000000

### Performance Patterns – Key Findings
Analysis shows that employee performance varies significantly across departments, education levels, and age groups. Mid-career employees and those with higher training scores tend to meet KPIs more consistently. Departments with focused training programs demonstrate better overall performance outcomes.


### EDA – Section 3.2 - Retention-Related Trend Analysis

In [50]:
#3.2.1 Tenure distribution (who stays longer?)
df['length_of_service'].describe()

count    17415.000000
mean         5.801493
std          4.175632
min          1.000000
25%          3.000000
50%          5.000000
75%          7.000000
max         34.000000
Name: length_of_service, dtype: float64

In [51]:
# A right-skewed tenure distribution suggests higher churn among early-career employees.

In [53]:
# 3.2.2 Tenure by performance
df.groupby('KPIs_met_more_than_80')['length_of_service'].mean().reset_index()

KPIs_met_more_than_80  length_of_service
0                      0           6.049978
1                      1           5.357600

In [54]:
# Employees meeting KPIs tend to have longer tenure, indicating a strong link between performance and retention.

In [55]:
# 3.2.3 Tenure by previous year rating
df.groupby('previous_year_rating')['length_of_service'].mean().reset_index()

previous_year_rating  length_of_service
0                   1.0           6.199595
1                   2.0           6.075851
2                   3.0           5.212653
3                   4.0           6.218631
4                   5.0           6.264868

In [56]:
# 3.2.4 Impact of awards on retention
df.groupby('awards_won')['length_of_service'].mean().reset_index()

awards_won  length_of_service
0           0           5.826846
1           1           4.742015

In [57]:
# Employees who received awards generally show higher average tenure, highlighting recognition as a retention driver.

In [58]:
# 3.2.5 Retention by age group

In [60]:
df.groupby('age_group')['length_of_service'].mean().reset_index()

C:\Users\jaffe\AppData\Local\Temp\ipykernel_9268\930077273.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  df.groupby('age_group')['length_of_service'].mean().reset_index()


age_group  length_of_service
0     20–30           3.015837
1     31–40           5.973854
2     41–50           9.373080
3     51–60          12.177419

In [61]:
# Mid-career age groups exhibit longer tenure compared to younger employees, suggesting early-career attrition risk.

### Retention Trends – Key Findings
Retention analysis using tenure as a proxy indicates that employees with higher performance ratings, KPI achievement, and awards tend to remain longer in the organization. Early-career employees show shorter tenure, suggesting higher attrition risk during initial years.

In [62]:
# EDA – Section 3.3 - Relationship Analysis → Actionable Insights

In [64]:
# 3.3.1 Training score vs KPI achievement
df.groupby('KPIs_met_more_than_80')['avg_training_score'].mean().reset_index()


KPIs_met_more_than_80  avg_training_score
0                      0           62.456158
1                      1           64.466720

In [65]:
# Employees meeting KPIs consistently have higher training scores, indicating training effectiveness.

In [68]:
# 3.3.2 Training count vs performance
df.groupby('no_of_trainings')[['avg_training_score',
                               'KPIs_met_more_than_80']].mean().reset_index()

no_of_trainings  avg_training_score  KPIs_met_more_than_80
0                1           62.789530               0.367736
1                2           65.026256               0.333333
2                3           64.220779               0.306122
3                4           62.678832               0.204380
4                5           66.576923               0.307692
5                6           62.250000               0.000000
6                7           63.666667               0.166667
7                8           71.000000               0.500000
8                9           70.000000               0.500000

In [69]:
# 3.3.3 Department-wise retention risk
df.groupby('department')['length_of_service'].mean().reset_index()


department  length_of_service
0          Analytics           4.998232
1            Finance           5.009975
2                 HR           5.629362
3              Legal           4.500000
4         Operations           6.429342
5        Procurement           6.187500
6                R&D           4.798193
7  Sales & Marketing           5.753023
8         Technology           5.844020

In [70]:
# Certain departments show consistently lower average tenure, indicating higher retention risk.

In [71]:
# 3.3.4 Combined performance & retention view
df.groupby('department')[['KPIs_met_more_than_80',
                           'avg_training_score',
                           'length_of_service']].mean().reset_index()


department  KPIs_met_more_than_80  avg_training_score  \
0          Analytics               0.400118           84.565115   
1            Finance               0.397756           60.326683   
2                 HR               0.361011           50.386282   
3              Legal               0.355422           59.533133   
4         Operations               0.440692           60.351589   
5        Procurement               0.373214           70.180357   
6                R&D               0.448795           84.451807   
7  Sales & Marketing               0.277208           50.056614   
8         Technology               0.356071           79.847658   

   length_of_service  
0           4.998232  
1           5.009975  
2           5.629362  
3           4.500000  
4           6.429342  
5           6.187500  
6           4.798193  
7           5.753023  
8           5.844020

### Key Relationships and Insights
The analysis reveals a strong relationship between training effectiveness, performance outcomes, and employee retention. Departments with higher training scores and KPI achievement demonstrate longer employee tenure, emphasizing the importance of targeted training and recognition programs.

# Step 4: Recommendations & Conclusion

## 4. Recommendations

Based on the exploratory data analysis, the following recommendations are proposed to improve employee productivity and reduce turnover:

1. **Strengthen Targeted Training Programs**  
   Employees with higher average training scores consistently met KPIs and demonstrated longer tenure. The organization should focus on improving training quality and tailoring programs by department to enhance performance outcomes.

2. **Focus on Early-Career Employee Retention**  
   Employees with shorter tenure and lower age groups exhibited higher attrition risk. Structured onboarding programs, mentorship initiatives, and early performance support can help improve retention among early-career employees.

3. **Leverage Performance Ratings for Retention Strategy**  
   Employees with higher previous-year ratings tended to stay longer. Regular performance feedback, transparent appraisal systems, and growth-oriented discussions can improve engagement and retention.

4. **Expand Recognition and Rewards Programs**  
   Award-winning employees showed higher average tenure, indicating that recognition positively impacts retention. Expanding reward and recognition initiatives can motivate employees and reduce turnover.

5. **Department-Specific Interventions**  
   Certain departments demonstrated lower average tenure and performance. Customized engagement strategies, leadership support, and department-level training interventions should be implemented to address these gaps.


## 5. Conclusion

This HR analytics project analyzed employee performance and retention-related trends using demographic, training, and performance data. The findings highlight a strong relationship between training effectiveness, performance outcomes, and employee tenure. Employees who achieved KPIs, received awards, and demonstrated strong training performance tended to remain longer within the organization.

The analysis also identified early-career employees and specific departments as higher retention risk areas, emphasizing the need for targeted engagement and support strategies. Overall, the insights derived from this analysis can help HR teams design data-driven interventions to improve employee productivity, engagement, and long-term retention.
